src/app/AutoResubscribeClient.cpp

```diff
@@ -0,0 +1,107 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <app/AutoResubscribeClient.h>
+#include <app/InteractionModelEngine.h>
+
+namespace chip {
+namespace app {
+
+CHIP_ERROR AutoResubscribeClient::Init(Messaging::ExchangeManager * apExchangeMgr, ReadClient::Callback * apCallback,
+                                       ResubscribePolicyDelegate * apResubscribeDelegate)
+{
+    mpExchangeMgr = apExchangeMgr;
+    mCallback.Init(apCallback, this);
+    if (apResubscribeDelegate == nullptr)
+    {
+        mpResubscribePolicyDelegate = &mDefaultResubscribeImpl;
+    }
+    else
+    {
+        mpResubscribePolicyDelegate = apResubscribeDelegate;
+    }
+    app::InteractionModelEngine * engine = app::InteractionModelEngine::GetInstance();
+    return engine->NewReadClient(&mpReadClient, ReadClient::InteractionType::Subscribe, &mCallback);
```

Comment:
@bzbarsky-apple  currently we still need to get read client from client pool for subscription behavior, the reason is as below, during post-subscription, we need to use InteractionModelEngine to process incoming unsolicited report in initiator side, where we would check read client pool and check if that is matched one. https://github.com/project-chip/connectedhomeip/blob/master/src/app/InteractionModelEngine.cpp#L395. I think we would optimize this client pool in https://github.com/project-chip/connectedhomeip/pull/13108, where we would use heap to allocated this pool and list to manage it.